python/thirdweb-ai/src/thirdweb_ai/services/storage.py

```diff
@@ -0,0 +1,176 @@
+        """Post files to a URL using async client with proper authorization headers."""
+        headers = self._make_headers()
+        # Remove the Content-Type as httpx will set it correctly for multipart/form-data
+        headers.pop("Content-Type", None)
+
+        async with httpx.AsyncClient() as client:
+            response = await client.post(url, files=files, headers=headers)
+            response.raise_for_status()
+            return response.json()
+
+    def _is_json_serializable(self, data: Any) -> bool:
+        """Check if data is JSON serializable (dict, dataclass, or BaseModel)."""
+        return isinstance(data, dict) or is_dataclass(data) or isinstance(data, BaseModel)
+
+    def _convert_to_json(self, data: Any) -> str:
+        """Convert data to JSON string."""
+        if isinstance(data, dict):
+            return json.dumps(data)
+        if is_dataclass(data):
+            # Handle dataclass properly
+            if isinstance(data, type):
+                raise ValueError(f"Expected dataclass instance, got dataclass type: {data}")
+            return json.dumps(asdict(data))
+        if isinstance(data, BaseModel):
+            return data.model_dump_json()
+        raise ValueError(f"Cannot convert {type(data)} to JSON")
+
+    def _is_valid_path(self, path: str) -> bool:
+        """Check if the string is a valid file or directory path."""
+        return Path(path).exists()
+
+    async def _prepare_directory_files(
+        self, directory_path: Path, chunk_size: int = 8192
+    ) -> list[tuple[str, BytesIO, str]]:
+        """
+        Prepare files from a directory for upload, preserving directory structure.
+        Returns a list of tuples (relative_path, file_buffer, content_type).
+        """
+        files_data = []
+
+        for root, _, files in os.walk(directory_path):
+            for file in files:
+                file_path = Path(root) / file
+                # Preserve the directory structure in the relative path
+                relative_path = str(file_path.relative_to(directory_path))
+                content_type = mimetypes.guess_type(str(file_path))[0] or "application/octet-stream"
+
+                # Create a buffer and read the file in chunks
+                buffer = BytesIO()
+                async for chunk in async_read_file_chunks(file_path, chunk_size):
+                    buffer.write(chunk)
+                buffer.seek(0)  # Reset buffer position
+
+                files_data.append((relative_path, buffer, content_type))
+
+        return files_data
+
+    @tool(
+        description="Upload a file, directory, or JSON data to IPFS. Stores any type on decentralized storage and returns an IPFS URI."
+    )
+    async def upload_to_ipfs(
+        self,
+        data: Annotated[
+            Any, "Data to upload: can be a file path, directory path, dict, dataclass, or BaseModel instance."
+        ],
+        chunk_size: Annotated[int, "Size of chunks to read from file."] = 8192,
```

Comment:
don't think we need to expose it as a tool parameter?
we can just pick a good default